### PR TITLE
Problem: released goleveldb binaries has external libc dependencies

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #464
Solution:
- build linux and x86_64-darwin goleveldb release binaries using nix too.
- upload checksums for nix built binaries.

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: default.nix [hunk 1]
```diff
@@ -47,4 +47,11 @@ buildGoApplication rec {
     -ldflags=
     ${ldflags}
   '';
+
+  meta = with lib; {
+    description = "Official implementation of the Cronos blockchain protocol";
+    homepage = "https://cronos.org/";
+    license = licenses.asl20;
```
Comment (yihuang) [inline default.nix:54]: https://github.com/NixOS/nixpkgs/blob/master/lib/licenses.nix#L95
Apache License 2.0

File: flake.nix [hunk 2]
```diff
@@ -59,12 +57,11 @@
...
         bundle-exe = import nix-bundle-exe { pkgs = final; };
-        bundle-exe-tarball = drv:
-          let bundle = final.bundle-exe drv;
-          in
-          final.runCommand bundle.name { } ''
-            "${final.gnutar}/bin/tar" cfzhv $out -C ${bundle} .
-          '';
+        make-tarball = drv: with final; runCommand drv.name { } ''
+          "${gnutar}/bin/tar" cfv - -C ${drv} \
+            --owner=0 --group=0 --mode=u+rw,uga+r --hard-dereference . \
```
Comment (yihuang) [inline flake.nix:62]: don't follow symbolic links.
don't leak the building user in the file attributes.
keep the normal file permissions.
follow hard links.

File: flake.nix [hunk 3]
```diff
@@ -59,12 +57,11 @@
...
+            | "${gzip}/bin/gzip -9 $out"
```
Comment (yihuang) [inline flake.nix:63]: keep gzip dependency explicit for better reproducibility.

File: default.nix [hunk 4]
```diff
@@ -10,7 +10,7 @@ let
   version = "v0.7.0";
   pname = "cronosd";
   tags = lib.concatStringsSep "," (
-    [ "ledger" "cgo" network ]
+    [ "ledger" "netgo" network ]
```
Comment (yihuang) [inline default.nix:13]: see: https://github.com/3noch/nix-bundle-exe/issues/14

File: flake.lock [hunk 5]
```diff
@@ -39,15 +39,16 @@
     "nix-bundle-exe": {
       "flake": false,
       "locked": {
-        "lastModified": 1651240934,
-        "narHash": "sha256-oLM2nqdxzCqNTxGtgFjIltOy5fU8iRiurKMBRExbUIY=",
-        "owner": "3noch",
+        "lastModified": 1652771427,
+        "narHash": "sha256-3Fv6+lGbxOVTRU00aE1SPIeRKgcxdFx7KgKT+xueSB0=",
+        "owner": "yihuang",
```
Comment (yihuang) [inline flake.lock:44]: TODO, update after upstream PR merged.

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -29,20 +29,36 @@ jobs:
...
+          cp result cronos_${GITHUB_REF_NAME:1}_${PLATFORM}.tar.gz
+
+          nix build -L ${FLAKE}#cronosd-testnet-$BUILD_TYPE
           cp result cronos_${GITHUB_REF_NAME:1}-testnet-rocksdb_${PLATFORM}.tar.gz
+
+          nix build -L ${FLAKE}#cronosd-testnet-goleveldb-$BUILD_TYPE
+          cp result cronos_${GITHUB_REF_NAME:1}-testnet_${PLATFORM}.tar.gz
+
+          nix-env -i coreutils -f '<nixpkgs>'
+          sha256sum *.tar.gz > checksums-x86-64.txt
```
Comment (yihuang) [inline .github/workflows/release.yml:57]:
```suggestion
          sha256sum *.tar.gz > "checksums-$PLATFORM.txt"
```

File: .goreleaser.yml [hunk 7]
```diff
@@ -161,35 +71,27 @@ archives:
...
   builds:
-    - cronosd-darwin-testnet
     - cronosd-darwin-arm64-testnet
     - cronosd-windows-testnet
-    - cronosd-linux-testnet
-    - cronosd-linux-arm64-testnet
-  
+ 
 checksum:
   name_template: 'checksums.txt'
```
Comment (yihuang) [inline .goreleaser.yml:96]:
```suggestion
  name_template: 'checksums-others.txt'
```

File: flake.nix [hunk 8]
```diff
@@ -59,12 +57,11 @@
...
-        bundle-exe-tarball = drv:
-          let bundle = final.bundle-exe drv;
-          in
-          final.runCommand bundle.name { } ''
-            "${final.gnutar}/bin/tar" cfzhv $out -C ${bundle} .
-          '';
+        make-tarball = drv: with final; runCommand drv.name { } ''
+          "${gnutar}/bin/tar" cfv - -C ${drv} \
+            --owner=0 --group=0 --mode=u+rw,uga+r --hard-dereference . \
+            | "${gzip}/bin/gzip" -9 > $out
```
Comment (yihuang) [inline flake.nix:63]: ref the nix gzip explicitly for better reproducibility.

File: flake.nix (hunk 2 above)
Comment (tomtau) [reply flake.nix:62]: maybe good to put this as a comment directly in the source code?

**Comment by yihuang** [comment]
https://github.com/yihuang/cronos/releases/tag/v0.0.38-test
test run result.

File: flake.nix (hunk 2 above)
Comment (yihuang) [reply flake.nix:64]: sounds good, done.

**Review by tomtau — APPROVED** [review]
should this undergo QA testing that e.g. ledger / usb is still working with these new bundled binaries on different platforms?

**Comment by yihuang** [comment]
> should this undergo QA testing that e.g. ledger / usb is still working with these new bundled binaries on different platforms?

sounds good. loop in @macong-cdc

**Comment by tomtau** [comment]
oh ok, in that case, one can keep using the old binary for CLI

**Comment by yihuang** [comment]
> oh ok, in that case, one can keep using the old binary for CLI

I'll merge this for now, if the test failed, we'll fix in another PR.

**Pull request merged**